### DISCUSSION TITLE
GoPro Exit Macro

### DISCUSSION THREAD

**Discussion by sT0v** [gopro/labs#1198] [Q&A] [answered]
#### Goal:
Exit macros on camera mode change without having to cycle power.

```
=Mv==vM"$M $v ok"!R3!X

=Mv        \\ load starting camera mode
==vM       \\ check if current mode is equal 
"$M $v ok" \\ log message 
!R3        \\ repeat from equality check 
!X         \\ assuming if the equality check fails then stop 

```

#### Issue:
When I test this and change the camera mode, the logging message stops as expected but `Labs Processing` still remains.

Is there a better build in way to do this? Otherwise, what is missing in my script? Is there an inequality check?

**Comment by dnewman-gpsw** [comment, marked as answer]
You script will next get to the exit (!X)

This is what you have written:
```
=Mv
loop:
if(v == M)
   print("$M $v ok")
goto loop:
exit();
```

As I don't know the intended purpose of the script, I don't have a fix.  Maybe you meant: =Mv==vM"$M $v ok"+!R3!X

```
=Mv
loop:
if(v == M)
{  
   print("$M $v ok")
   goto loop:
}
exit();
```

**Reply by sT0v** [reply]
> As I don't know the intended purpose of the script, I don't have a fix. Maybe you meant: =Mv==vM"$M $v ok"+!R3!X

The purpose is to exit a script when I change camera modes.

What does the `+` mean here?

**Reply by dnewman-gpsw** [reply]
The Plus includes the next function inside the if condition.  This is to make up for the QR programming language not having braces like C or indenting like Python.

**Reply by sT0v** [reply]
> The Plus includes the next function inside the if condition.  This is to make up for the QR programming language not having braces like C or indenting like Python.

Perfect. I'd been wanting to know that for a bit now thank you.


So the script works as intended now, but I can't seem to extend this into a macro.

QR1:`$EXIT="=Mv==vM'$M $v ok'+!R3!X"`
QR2: `^EXIT`

Is this possible?

**Reply by dnewman-gpsw** [reply]
Macros have to be permanent. Try *EXIT="=Mv==vM'$M $v ok'+!R3!X"

**Reply by dnewman-gpsw** [reply]
That works here.

**Reply by sT0v** [reply]
> Macros have to be permanent. Try *EXIT="=Mv==vM'$M $v ok'+!R3!X"

**When I include this code in the script for the macro, will the `!R3` go-to function need to be amended?**

If so, is there anyway that could be dynamic?

Intended use would be:
Run script A `RUNA` if condition A `CNDA` until I switch the camera mode so that I can kill it by hand.

```python
while cond_a:
   run_a()
exit()
```

`<CNDA^RUNA~EXIT`

**Reply by dnewman-gpsw** [reply]
The !R3 should work within the macro, jus to character pos three within the macro.  However if you have issue, there is also relative jumps with !R-13 (jump 13 characters back form the !)

Scan this: 
```
"Hello""World $l"<l10!R-14!X
```

Maybe useful.

**Reply by sT0v** [reply]
In your example, I would expect it to print "Hello" once and "World" the 10 times.

Since it should jump to the opening quote for World character `"`. Why is it printing Hello each time as well?

`"Hello""World $l"<l10!R7!X`
This is the behavior I'd like using your negative indexed goto loop idea.

**Reply by dnewman-gpsw** [reply]
It works on HERO12 and 13 only so far. The negative offset was fairly new.

**Reply by sT0v** [reply]
Ok, yeah I'm on the 11, otherwise, that's a sweet solution.

**Reply by dnewman-gpsw** [reply]
It should come to the 11, maybe early in the new year.
